### PR TITLE
[C++ API] Make error message for empty module friendlier

### PR DESCRIPTION
**Pull request by goldsborough**
In our pimpl system, default constructing a module holder default constructs the contained module. This means `Linear linear;` is ill-formed, since `Linear` doesn't have a default constructor. Instead we require `Linear linear = nullptr;` to get the empty state of the `Linear`. This PR makes the error message for the ill-formed case nicer.

I had to change the forwarding constructors of most of our modules for this, but that's a minor adjustment.

E.g.

```
Linear linear;

In file included from /home/psag/pytorch/pytorch/torch/csrc/api/include/torch/nn/module.h:5:0,
                 from /home/psag/pytorch/pytorch/test/cpp/api/module.cpp:3:
/home/psag/pytorch/pytorch/torch/csrc/api/include/torch/nn/pimpl.h: In instantiation of ‘torch::nn::ModuleHolder<Contained>::ModuleHolder() [with Contained = torch::nn::LinearImpl]’:
/home/psag/pytorch/pytorch/torch/csrc/api/include/torch/nn/modules/dropout.h:45:1:   required from here
/home/psag/pytorch/pytorch/torch/csrc/api/include/torch/nn/pimpl.h:46:5: error: static assertion failed: You are trying to default construct a module which has no default constructor. Use = nullptr to give it the empty state (like an empt
y std::shared_ptr).
     static_assert(
```

@ebetica @ezyang

### CODE REVIEW

File: torch/csrc/api/include/torch/nn/pimpl.h [hunk 1]
```diff
@@ -38,10 +38,22 @@ class ModuleHolder : torch::detail::ModuleHolderIndicator {
...
+  /// Default constructs the contained module if if has a default constructor,
+  /// else produces a static error. NOTE: This uses the behavior of template
+  /// classes in C++ that constructors (or any methods) are only compiled when
+  /// actually used.
+  ModuleHolder()  {
+    static_assert(
+        std::is_default_constructible<Contained>::value,
+        "You are trying to default construct a module which has "
+        "no default constructor. Use = nullptr to give it the empty state "
+        "(like an empty std::shared_ptr).");
```
Comment (ebetica) [inline torch/csrc/api/include/torch/nn/pimpl.h]: nit: provide an example maybe:
```
"(do `Linear x = nullptr;` instead of `Linear x;`)"
```

**Review by ebetica — APPROVED** [review]

**Review by ezyang — APPROVED** [review]
I'll note that C++ newbies never read `static_assert` error messages, but there's not really much we can do here...

**Comment by ebetica** [comment]
@ezyang Well, the previous error message did not even have the line pointing to the exact error, so I consider this a big win :D

**Pull request closed without merging**